### PR TITLE
python3Packages.minichain: init at 0.3.3

### PR DESCRIPTION
**Pull request by samuela**
## Description of changes

Package https://github.com/srush/MiniChain.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/minichain/default.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+, lib
+, manifest-ml
+, openai
+, pytestCheckHook
+, pythonRelaxDepsHook
+}:
+
+buildPythonPackage rec {
+  pname = "minichain";
+  version = "0.3.3";
```
Comment (natsukium) [inline pkgs/development/python-modules/minichain/default.nix:15]:
```suggestion
  version = "0.3.3";
  format = "setuptools";
```
It seems to be preferable to set `format = "setuptools";` explicitly these days.

File: pkgs/development/python-modules/minichain/default.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    "datasets"
+    "faiss-cpu"
+    "jinja2-highlight"
+    "trio"
+
+    # Not used anywhere:
+    "eliot-tree"
+
+    # Not yet packaged in nixpkgs:
+    "gradio"
```
Comment (natsukium) [inline pkgs/development/python-modules/minichain/default.nix:42]: FYI: gradio has recently been added to nixpkgs-staging in https://github.com/NixOS/nixpkgs/pull/241558.
I expect we'll be able to add it to the "optional" dependencies soon.

**Comment by natsukium** [comment]
@ofborg build python310Packages.minichain

File: pkgs/development/python-modules/minichain/default.nix (hunk 2 above)
Comment (samuela) [reply pkgs/development/python-modules/minichain/default.nix:43]: Cool, didn't know that! I'll make a mental note that it can be added in a future PR

**Comment by natsukium** [comment]
```
       > Executing pytestCheckPhase
       > ============================= test session starts ==============================
       > platform linux -- Python 3.10.12, pytest-7.4.0, pluggy-1.2.0
       > rootdir: /build/minichain-0.3.3
       > plugins: devtools-0.11.0, anyio-3.7.0
       > collected 0 items                                                              
       >
       > ============================ no tests ran in 0.01s =============================
       > /nix/store/rk3g4mxrs84bv1vanqnzpdkhm73ksr0z-stdenv-linux/setup: line 1596: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/cmbin6brgf1y8gahij5bry4g0h1fqv4r-python3.10-minichain-0.3.3.drv'.
```
The source fetched from PyPI may not contain any tests.

**Comment by natsukium** [comment]
Ah, yes, could you please add `doCheck = false`?

**Comment by samuela** [comment]
@natsukium  done in https://github.com/NixOS/nixpkgs/pull/248195/commits/cb1bfe97d3d8dfe4c1ca60b28dba20870683e923

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 248195` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.minichain</li>
    <li>python311Packages.minichain.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.minichain</li>
    <li>python310Packages.minichain.dist</li>
  </ul>
</details>

**Comment by natsukium** [comment]
It doesn't seem to support python3.11.
Can we `disabled = pythonAtLeast "3.11"`?
Besides this, LGTM.
```
Check whether the following modules can be imported: minichain
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <lambda>
  File "/nix/store/h29rxipmgnk10p9hx92ndkmfrh26p9b5-python3-3.11.4/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/nix/store/jjc5nvqzxbfsharffs2g0mlwp4fs591y-python3.11-minichain-0.3.3/lib/python3.11/site-packages/minichain/__init__.py", line 1, in <module>
    from .backend import (
  File "/nix/store/jjc5nvqzxbfsharffs2g0mlwp4fs591y-python3.11-minichain-0.3.3/lib/python3.11/site-packages/minichain/backend.py", line 309, in <module>
    @dataclass
     ^^^^^^^^^
  File "/nix/store/h29rxipmgnk10p9hx92ndkmfrh26p9b5-python3-3.11.4/lib/python3.11/dataclasses.py", line 1230, in dataclass
    return wrap(cls)
           ^^^^^^^^^
  File "/nix/store/h29rxipmgnk10p9hx92ndkmfrh26p9b5-python3-3.11.4/lib/python3.11/dataclasses.py", line 1220, in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/h29rxipmgnk10p9hx92ndkmfrh26p9b5-python3-3.11.4/lib/python3.11/dataclasses.py", line 958, in _process_class
    cls_fields.append(_get_field(cls, name, type, kw_only))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/h29rxipmgnk10p9hx92ndkmfrh26p9b5-python3-3.11.4/lib/python3.11/dataclasses.py", line 815, in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
ValueError: mutable default <class 'minichain.backend.RunLog'> for field run_log is not allowed: use default_factory
```

**Comment by samuela** [comment]
@natsukium oh good catch, fixed in https://github.com/NixOS/nixpkgs/pull/248195/commits/9d13a1dc98690f55df5a78ed671b75f7f56ee532

**Review by natsukium — APPROVED** [review]
LGTM, thanks!

**Pull request merged**